zerver/actions/user_settings.py

```diff
@@ -576,25 +576,51 @@ def do_change_user_setting(
             status = UserPresence.LEGACY_STATUS_ACTIVE_INT
             presence_time = timezone_now()
         else:
-            # HACK: Remove existing presence data for the current user
-            # when disabling presence. This hack will go away when we
-            # replace our presence data structure with a simpler model
-            # that doesn't separate individual clients.
-            UserPresence.objects.filter(user_profile_id=user_profile.id).delete()
-
-            # We create a single presence entry for the user, old
-            # enough to be guaranteed to be treated as offline by
-            # correct clients, such that the user will, for as long as
-            # presence remains disabled, appear to have been last
-            # online a few minutes before they disabled presence.
+            # We want to ensure the user's presence data is such that they will be treated as offline by
+            # correct clients,
+            # There are two cases:
             #
-            # We add a small additional offset as a fudge factor in
-            # case of clock skew.
-            status = UserPresence.LEGACY_STATUS_IDLE_INT
-            presence_time = timezone_now() - timedelta(
+            # (1) If the user has a recent presence, we backdate it so that the user will, for as long as
+            #     presence remains disabled, appear to have been last online a few minutes
+            #     before they disabled presence.
+            # (2) If the user only has a presence older than what our backdate would be, we keep the original
+            #     value - it already guarantees that the user will appear to be offline.
+            backdated_presence_time = timezone_now() - timedelta(
+                # We add a small additional offset as a fudge factor in
+                # case of clock skew.
                 seconds=settings.OFFLINE_THRESHOLD_SECS + 120
             )
 
+            try:
+                presence: UserPresence | None = UserPresence.objects.get(user_profile=user_profile)
+                assert (
+                    presence is not None
+                )  # mypy doesn't understand that presence cannot be None here.
+                assert presence.last_connected_time is not None
+                if presence.last_connected_time <= backdated_presence_time:
+                    # This is case (2), so we don't need to do anything. presence is already as intended.
+                    return
+
+                presence_time = backdated_presence_time
```

Comment:
Thinking about this a bit more -- should this be `min(backdated_presence_time, presence.last_active_time)`? I think that'd be best for avoiding perturbing your last active time in the event that the existing time was a few days ago (not very likely; since logging in updates presence, I'm not sure it'll ever really be in that state at the moment one enables invisible mode).

Reply:
I guess a thing we could do is have:

(1) The update event be `STATUS_IDLE` rather than active
(2) Replace the `presence.delete` with writing to the database to make `last_active_time` equal `min(backdated_presence_time - 1), last_active_time` and `last_connected_time=min(backdated_presence_time - 1), last_active_time - 1)`